File: python/blog/ds/linkedList.py
from node import Node

class LinkedList:
    head = None
    tail = None
    length = 0

    def toString(self):
        tmp = self.head
        result = []

        while(tmp is not None):
            result.append(str(tmp.data))
            tmp = tmp.next

        return ','.join(result)

    def addFirst(self, input):
        node = Node(input)

        if self.head is not None:
            node.next = self.head
            
        if self.tail is None:
            self.tail = node
        
        self.head = node
        self.length += 1

    def addLast(self, input):
        node = Node(input)

        if self.tail is not None:
            self.tail.next = node
        
        if self.head is None:
            self.head = node

        self.tail = node
        self.length += 1

    def _getNode(self, idx):
        tmp = self.head
        for i in range(idx):
            if tmp is not None:
                tmp = tmp.next
        return tmp

    def addFront(self, input, idx):
        if idx == 0:
            self.addFirst(input)
        elif idx >= self.length:
            self.addLast(input)
        else:
            prev = self._getNode(idx - 1)

            node = Node(input)
            node.next = prev.next
            prev.next = node
            self.length += 1

    def addBack(self, input, idx):
        if idx >= self.length:
            self.addLast(input)
        else:
            current = self._getNode(idx)
            
            node = Node(input)
            node.next = current.next
            current.next = node

            if (node.next is None):
                self.tail = node

            self.length += 1

    def removeFirst(self):
        if self.empty():
            return None

        deleteNode = self.head
        result = deleteNode.data
        self.head = deleteNode.next

        if self.head is None:
            self.tail = None
        
        deleteNode = None
        self.length -= 1

        #print('length: ', self.length, ', head: ', self._getNodeValue(self.head), ', tail:', self._getNodeValue(self.tail))

        return result

    def remove(self, idx):
        if idx == 0:
            return self.removeFirst()
        elif idx >= self.length:
            return None
        else:
            prev = self._getNode(idx - 1)
            delete = prev.next
            result = delete.data

            prev.next = delete.next

            if prev.next is None:
                self.tail = prev
            
            delete = None
            self.length -= 1

            #print('length: ', self.length, ', head: ', self._getNodeValue(self.head), ', tail:', self._getNodeValue(self.tail))
            
            return result

    def removeLast(self):
        if self.empty():
            return None
        return self.remove(self.length - 1)

    def size(self):
        return self.length

    def empty(self):
        return self.head is None

    def get(self, idx):
        if idx < 0 or idx >= self.length:
            return None

        return self._getNode(idx).data

    def indexOf(self, data):
        tmp = self.head
        for i in range(self.length):
            if tmp.data == data:
                return i
            tmp = tmp.next
        return -1

    def _getNodeValue(self, node):
        if node is None:
            return None
        return node.data

File: c++/blog/ds/linkedlist.cpp
#include <iostream>

using namespace std;

class Node {
	public:
		int data;
		Node* next;
		Node(int, Node*);
};

class LinkedList {
	private:
		Node* head = NULL;
		Node* tail = NULL;
		int length = 0;
		Node* getNode(int);
	public:
		void toString();
		void tearDown();

		void addFirst(int);
		void addLast(int);
		void addFront(int, int);
		void addBack(int, int);
		int removeFirst();
		int remove(int);
		int removeLast();
		bool empty();
		int size();
		int get(int);
		int indexOf(int);
};

Node::Node(int d, Node* n)
{
	this->data = d;
	this->next = n;
}

void LinkedList::toString()
{
	Node* tmp = this->head;
	for (int i = 0; i < length; i++) {
		cout << tmp->data << ", ";
		tmp = tmp->next;
	}
	cout << endl;
}

void LinkedList::tearDown()
{
	Node* tmp = this->head;
	Node* next;
	for (int i = 0; i < length; i++) {
		next = tmp->next;
		delete tmp;
		tmp = next;
	}
	this->length = 0;
	this->head = NULL;
	this->tail = NULL;
}

void LinkedList::addFirst(int input)
{
	Node* node = new Node(input, NULL);

	if (this->head != NULL)
		node->next = this->head;

	if (this->tail == NULL)
		this->tail = node;

	this->head = node;
	this->length++;
}

void LinkedList::addLast(int input)
{
	Node* node = new Node(input, NULL);

	if (this->tail != NULL)
		this->tail->next = node;

	if (this->head == NULL)
		this->head = node;

	this->tail = node;
	this->length++;
}

Node* LinkedList::getNode(int idx)
{
	Node* tmp = this->head;
	for (int i = 0; i < idx; i++) {
		if (tmp->next != NULL) {
			tmp = tmp->next;
		}
	}

	return tmp;
}

void LinkedList::addFront(int input, int idx)
{
	if (idx == 0) {
		this->addFirst(input);
	} else if (idx >= this->length) {
		this->addLast(input);
	} else {
		Node* prev = this->getNode(idx - 1);
		Node* node = new Node(input, NULL);

		node->next = prev->next;
		prev->next = node;
		this->length++;
	}
}

void LinkedList::addBack(int input, int idx)
{
	if (idx >= this->length) {
		this->addLast(input);
	} else {
		Node* current = this->getNode(idx);
		Node* node = new Node(input, NULL);

		node->next = current->next;
		current->next = node;

		if (node->next == NULL)
			this->tail = node;

		this->length++;
	}
}

int LinkedList::removeFirst()
{
	if (this->empty()) {
		return -1;
	} else {
		Node* del = this->head;
		int result = del->data;

		this->head = del->next;
		if (this->head == NULL)
			this->tail = NULL;

		delete del;
		this->length--;

		return result;
	}
}

int LinkedList::remove(int idx)
{
	if (idx == 0) {
		return this->removeFirst();
	} else if (idx >= this->length) {
		return -1;
	} else {
		Node* prev = this->getNode(idx - 1);
		Node* del = prev->next;
		int result = del->data;

		prev->next = del->next;
		if (prev->next == NULL)
			this->tail = prev;

		delete del;
		this->length--;

		return result;
	}
}

int LinkedList::removeLast()
{
	if (empty())
		return -1;
	return this->remove(this->length - 1);
}

bool LinkedList::empty()
{
	return this->head == NULL;
}

int LinkedList::size()
{
	return this->length;
}

int LinkedList::get(int idx)
{
	if (idx < 0 || idx >= this->length)
		return -1;
	return this->getNode(idx)->data;
}

int LinkedList::indexOf(int data)
{
	Node* tmp = this->head;
	for (int i = 0; i < this->length; i++) {
		if (tmp->data == data)
			return i;
		tmp = tmp->next;
	}
	return -1;
}

void addFirstTest(LinkedList list)
{
	// add first - 4,3,2,1
	list.addFirst(1);
	list.addFirst(2);
	list.addFirst(3);
	list.addFirst(4);
	cout << "add first: ";
	list.toString();

	list.tearDown();
}

void addLastTest(LinkedList list)
{
	// add last - 1,2,3,4
	list.addLast(1);
	list.addLast(2);
	list.addLast(3);
	list.addLast(4);
	cout << "add last: ";
	list.toString();

	list.tearDown();
}

void addFrontTest(LinkedList list)
{
	// add front - 1,3,4,2
	list.addFront(1, 0);
	list.addFront(2, 1);
	list.addFront(3, 1);
	list.addFront(4, 2);
	cout << "add front: ";
	list.toString();

	list.tearDown();
}

void addBackTest(LinkedList list)
{
	// add back - 1,2,4,3
	list.addBack(1, 0);
	list.addBack(2, 1);
	list.addBack(3, 1);
	list.addBack(4, 1);
	cout << "add back: ";
	list.toString();

	list.tearDown();
}

void removeFirstTest(LinkedList list)
{
	list.addLast(1);
	list.addLast(2);
	list.addLast(3);
	list.addLast(4);
	list.addLast(5);

	printf("remove first: %d ", list.removeFirst());
	printf("%d ", list.removeFirst());
	printf("%d ", list.removeFirst());
	printf("%d ", list.removeFirst());
	printf("%d ", list.removeFirst());
	printf("%d\n", list.removeFirst());

	list.tearDown();
}

void removeTest(LinkedList list)
{
	list.addLast(1);
	list.addLast(2);
	list.addLast(3);
	list.addLast(4);
	list.addLast(5);

	printf("remove: %d ", list.remove(5));
	printf("%d ", list.remove(4));
	printf("%d ", list.remove(0));
	printf("%d ", list.remove(1));
	printf("%d ", list.remove(1));
	printf("%d\n", list.remove(0));

	list.tearDown();
}

void removeLastTest(LinkedList list)
{
	list.addLast(1);
	list.addLast(2);
	list.addLast(3);
	list.addLast(4);
	list.addLast(5);

	printf("remove last: %d ", list.removeLast());
	printf("%d ", list.removeLast());
	printf("%d ", list.removeLast());
	printf("%d ", list.removeLast());
	printf("%d ", list.removeLast());
	printf("%d\n", list.removeLast());

	list.tearDown();
}

void emptyTest(LinkedList list)
{
	printf("empty: %s ", list.empty()? "true" : "false");

	list.addLast(1);
	printf("%s\n", list.empty()? "true" : "false");

	list.tearDown();
}

void sizeTest(LinkedList list)
{
	printf("size: %d ", list.size());

	list.addLast(1);
	printf("%d\n", list.size());

	list.tearDown();
}

void getTest(LinkedList list)
{
	list.addLast(1);
	list.addLast(2);
	list.addLast(3);
	list.addLast(4);
	list.addLast(5);

	printf("get: %d ", list.get(-1));
	printf("%d ", list.get(0));
	printf("%d ", list.get(1));
	printf("%d ", list.get(2));
	printf("%d ", list.get(3));
	printf("%d ", list.get(4));
	printf("%d\n", list.get(5));

	list.tearDown();
}

void indexOfTest(LinkedList list)
{
	list.addLast(1);
	list.addLast(2);
	list.addLast(3);
	list.addLast(4);
	list.addLast(5);

	printf("index of: %d ", list.indexOf(1));
	printf("%d ", list.indexOf(2));
	printf("%d ", list.indexOf(3));
	printf("%d ", list.indexOf(4));
	printf("%d ", list.indexOf(5));
	printf("%d\n", list.indexOf(6));

	list.tearDown();
}

void main()
{
	LinkedList list;
	
        addFirstTest(list);
	addLastTest(list);
	addFrontTest(list);
	addBackTest(list);

	removeFirstTest(list);
	removeTest(list);
	removeLastTest(list);

	emptyTest(list);
	sizeTest(list);
	getTest(list);
	indexOfTest(list);
}
